packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -1073,43 +1083,52 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
           // This is a fatal error.
           didFatal = true;
           onUncaughtError(thrownValue);
-          break;
-        }
+        } else {
+          if (__DEV__) {
+            // Reset global debug state
+            // We assume this is defined in DEV
+            (resetCurrentlyProcessingQueue: any)();
+          }
 
-        if (__DEV__) {
-          // Reset global debug state
-          // We assume this is defined in DEV
-          (resetCurrentlyProcessingQueue: any)();
-        }
+          const failedUnitOfWork: Fiber = nextUnitOfWork;
+          if (__DEV__ && replayFailedUnitOfWorkWithInvokeGuardedCallback) {
+            replayUnitOfWork(failedUnitOfWork, thrownValue, isAsync);
+          }
 
-        if (__DEV__ && replayFailedUnitOfWorkWithInvokeGuardedCallback) {
-          const failedUnitOfWork = nextUnitOfWork;
-          replayUnitOfWork(failedUnitOfWork, thrownValue, isAsync);
-        }
+          // TODO: we already know this isn't true in some cases.
+          // At least this shows a nicer error message until we figure out the cause.
+          // https://github.com/facebook/react/issues/12449#issuecomment-386727431
+          invariant(
+            nextUnitOfWork !== null,
```

Comment:
This is an invariant in DEV-only code. But we *know* it already crashes (https://github.com/facebook/react/issues/12449#issuecomment-386727431).
Maybe @acdlite can think of a better fix but for now I'm making the existing bug explicit in the hope that somebody gives us a repro case.